mypy/checker.py

```diff
@@ -2979,14 +2979,14 @@ def visit_operator_assignment_stmt(self,
     def visit_assert_stmt(self, s: AssertStmt) -> None:
         self.expr_checker.accept(s.expr)
 
-        if s.msg is not None:
-            self.expr_checker.accept(s.msg)
-
         if isinstance(s.expr, TupleExpr) and len(s.expr.items) > 0:
             self.fail(message_registry.MALFORMED_ASSERT, s)
 
         # If this is asserting some isinstance check, bind that type in the following code
-        true_map, _ = self.find_isinstance_check(s.expr)
+        true_map, else_map = self.find_isinstance_check(s.expr)
+        if s.msg is not None:
+            with self.binder.frame_context(can_skip=False, fall_through=1):
+                self.expr_checker.analyze_cond_branch(else_map, s.msg, None)
```

Comment:
Hm, sorry for being annoying, but it looks like something is still wrong, now _two_ frames are being pushed, one manually, and one in the helper. Why can't you just use the helper here?

The `None` context is indeed right, because we have no expectations about what should be the assertion "message" (it can be arbitrary object, not just a string).

Reply:
> Why can't you just use the helper here?

Mostly because I misunderstood the binder and its `fall_through` argument